Converters/ArcGIS/Speckle.Converters.ArcGIS3/ToSpeckle/Raw/MultipatchFeatureToSpeckleConverter.cs

```diff
@@ -44,83 +42,127 @@ public IReadOnlyList<Base> Convert(ACG.Multipatch target)
-    for (int idx = 0; idx < target.PartCount; idx++)
+    for (int i = 0; i < target.PartCount; i++)
     {
-      // get the patch type to get the point arrangement in the mesh
+      // get the patch type to get the point arrangement
       // https://pro.arcgis.com/en/pro-app/latest/sdk/api-reference/topic27403.html
-      ACG.PatchType patchType = target.GetPatchType(idx);
+      ACG.PatchType patchType = target.GetPatchType(i);
 
-      if (patchType == ACG.PatchType.TriangleStrip)
-      {
-        SOG.Mesh multipatch = target.CompleteMultipatchTriangleStrip(allPoints, idx);
-        multipatch.units = _settingsStore.Current.SpeckleUnits;
-        converted.Add(multipatch);
-      }
-      else if (patchType == ACG.PatchType.Triangles)
+      // get the points in the patch
+      List<ACG.MapPoint> points = new();
+      int ptStartIndex = target.GetPatchStartPointIndex(i);
+      for (int ptIdx = ptStartIndex; ptIdx < ptStartIndex + target.GetPatchPointCount(i); ptIdx++)
       {
-        SOG.Mesh multipatch = target.CompleteMultipatchTriangles(allPoints, idx);
-        multipatch.units = _settingsStore.Current.SpeckleUnits;
-        converted.Add(multipatch);
+        points.Add(target.Points[ptIdx]);
       }
-      else if (patchType == ACG.PatchType.TriangleFan)
+
+      switch (patchType)
       {
-        SOG.Mesh multipatch = target.CompleteMultipatchTriangleFan(allPoints, idx);
-        multipatch.units = _settingsStore.Current.SpeckleUnits;
-        converted.Add(multipatch);
+        case ACG.PatchType.TriangleStrip:
+          SOG.Mesh triangleStripPatch = GetMeshFromTriangleStripPatch(points);
+          converted.Add(triangleStripPatch);
+          break;
+        case ACG.PatchType.Triangles:
+          SOG.Mesh trianglesPatch = GetMeshFromTrianglesPatch(points);
+          converted.Add(trianglesPatch);
+          break;
+        case ACG.PatchType.TriangleFan:
+          SOG.Mesh triangleFanPatch = GetMeshFromTriangleFanPatch(points);
+          converted.Add(triangleFanPatch);
+          break;
+        case ACG.PatchType.FirstRing:
+          SOG.Mesh firstRingPatch = GetMeshFromFirstRingPatch(points);
+          converted.Add(firstRingPatch);
+          break;
+
+        default:
+          throw new ValidationException($"{patchType} patch type is not supported");
       }
-      // in case of RingMultipatch - return PolygonGeometry3d
-      // the following Patch Parts cannot be pushed to external method, as they will possibly, add voids/rings to the same GisPolygon
-      else if (patchType == ACG.PatchType.FirstRing)
+    }
+    return converted;
+  }
+
+  private SOG.Mesh GetMeshFromTriangleStripPatch(List<ACG.MapPoint> points)
+  {
+    List<double> pointCoords = points.SelectMany(x => new List<double>() { x.X, x.Y, x.Z }).ToList();
+    List<int> faces = new();
+    List<double> vertices = new();
+
+    for (int i = 0; i < points.Count; i++)
+    {
+      if (i >= 2) // every new point adds a triangle
```

Comment:
this assumption smells a bit. i guess we rely on order of `points` here to create proper triangles. even if it ordered OK, i wonder what happens if points in non-convex shape, we might have overlapping triangles. (I know this was pre-existing😒)